tests/pos/i6247.scala

```diff
@@ -0,0 +1,10 @@
+implicit class NN[T](x:T|Null) {
+  def nn: T = x.asInstanceOf[T]
+}
+
+class Foo {
+  val x1: String|Null = null
+  x1.nn.length
+  val x2: String = x1.nn
+  x1.nn.length
```

Comment:
This line failed while the identical expression two lines above succeeded, so something fishy is going on, I've got what I think is a proper fix in https://github.com/lampepfl/dotty/pull/6301.

Reply:
Great! But I think it would be still an improvement to have the changes in this PR also. It is insurance against the fact that we are more lazy instantiating type variables than scalac. Do you agree?

Reply:
Without an example showing that this is needed I'm not sure, there's now been two actual bugs we caught that would have stayed hidden with this PR in: https://github.com/lampepfl/dotty/pull/6301 and https://github.com/lampepfl/dotty/pull/5484/commits/38c6d462da32462aea18df0be7b3d6be75c40da6. so I'd rather keep things as is for now.

Reply:
I agree. I just went through all the inference related // dotty deviation comments in the compiler codebase and could remove them all without this fix. Except one which was related to us dropping
weak conformance. So let's leave things as they are.